File: Sistema de Control Clínico Veterinario/Sistema_Control_Clinico_Veterinario/logica/Factura.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace SistemaVeterinario.logica
{
    public class Factura
    {
        public int IdFactura { get; set; }
        public DateTime Fecha { get; set; } = DateTime.Now;
        public Cliente Cliente { get; set; }
        public double Subtotal { get; set; }
        public double Impuestos { get; private set; }
        public double Total { get; private set; }
        public double Descuento { get; private set; }
        public string Estado { get; private set; } = "Pendiente";
        public List<Pago> Pagos { get; set; } = new List<Pago>();

        public Factura(Cliente cliente, double subtotal)
        {
            Cliente = cliente;
            Subtotal = subtotal;
            Recalcular();
        }

        private void Recalcular()
        {
            Impuestos = Subtotal * 0.16;
            Total = Subtotal + Impuestos - Descuento;
        }

        public void AplicarDescuento(double porcentaje)
        {
            Descuento = Subtotal * (porcentaje / 100.0);
            Recalcular();
        }

        public void AgregarPago(Pago p)
        {
            Pagos.Add(p);
            if (Pagos.Sum(x => x.Monto) >= Total)
                Estado = "Pagada";
        }

        public void Emitir()
        {
            Console.WriteLine("\n   ============ FACTURA ============");
            Console.WriteLine($"   No. Factura: {IdFactura}");
            Console.WriteLine($"   Cliente: {Cliente.Nombre}  NIT: {Cliente.Nit}");
            Console.WriteLine($"   Fecha: {Fecha:dd/MM/yyyy HH:mm}");
            Console.WriteLine($"   Subtotal:  Q{Subtotal:F2}");
            Console.WriteLine($"   Impuestos: Q{Impuestos:F2}");
            Console.WriteLine($"   Descuento: Q{Descuento:F2}");
            Console.WriteLine($"   TOTAL:     Q{Total:F2}");
            Console.WriteLine("   =================================");
        }
    }
}

File: Sistema de Control Clínico Veterinario/bluej/Factura.java
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class Factura {
    private int idFactura;
    private Date fecha;
    private double subtotal;
    private double impuestos;
    private double total;
    private String estado;
    private Cliente cliente;
    private List<Pago> pagos;
    
    public Factura(Cliente c, double subtotal) {
        this.cliente = c;
        this.subtotal = subtotal;
        this.impuestos = subtotal * 0.16;
        this.total = subtotal + impuestos;
        this.fecha = new Date();
        this.estado = "Pendiente";
        this.pagos = new ArrayList<>();
    }
    
    public double calcularTotal() { return total; }
    
    public void aplicarDescuento(double porcentaje) {
        this.total = total - (total * porcentaje / 100);
    }
    
    public void agregarPago(Pago p) {
        pagos.add(p);
        if (totalPagado() >= total) this.estado = "Pagada";
    }
    
    private double totalPagado() {
        double suma = 0;
        for (Pago p : pagos) suma += p.getMonto();
        return suma;
    }
    
    public String getEstado() { return estado; }
    public double getTotal() { return total; }
}